File: 조준장/구현/16234.py
import sys
from collections import deque


def bfs(a, b):
    dx = [0, 0, 1, -1]
    dy = [1, -1, 0, 0]
    q = deque([[a, b]])
    temp = [[a, b]]
    visited[a][b] = True
    while q:
        x, y = q.popleft()
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if 0 <= nx < n and 0 <= ny < n and visited[nx][ny] == 0:
                if l <= abs(graph[nx][ny] - graph[x][y]) <= r:
                    visited[nx][ny] = True
                    q.append([nx, ny])
                    temp.append([nx, ny])
    return temp


n, l, r = map(int, sys.stdin.readline().split())
graph = [list(map(int, sys.stdin.readline().split())) for _ in range(n)]
answer = 0

while True:
    country = []
    visited = [[False] * (n + 1) for _ in range(n + 1)]
    flag = False
    for i in range(n):
        for j in range(n):
            if not visited[i][j]:
                country = bfs(i, j)
                if len(country) > 1:
                    flag = True
                    population = (sum([graph[x][y] for x, y in country]) // len(country))

                    for x, y in country:
                        graph[x][y] = population
    if not flag:
        break
    answer += 1


print(answer)

File: 조준장/골드4/16234.kt
class `16234` {
    data class Node(val x: Int, val y: Int)

    lateinit var graph: List<MutableList<Int>>
    lateinit var visited: Array<BooleanArray>

    fun solve() {
        val br = System.`in`.bufferedReader()
        val bw = System.out.bufferedWriter()
        val sb = StringBuilder()

        val (n, l, r) = br.readLine().split(" ").map { it.toInt() }
        graph = List(n) { br.readLine().split(" ").map { it.toInt() }.toMutableList() }
        var answer = 0
        while (true) {
            var flag = false
            visited = Array(n + 1) { BooleanArray(n + 1) }

            for (i in 0 until n) {
                for (j in 0 until n) {
                    if (!visited[i][j]) {
                        val nodes = bfs(n, l, r, i, j)

                        if (nodes.size == 1) continue
                        flag = true

                        val num = nodes.sumOf { (x, y) -> graph[x][y] } / nodes.size
                        nodes.forEach { (x, y) ->
                            graph[x][y] = num
                        }
                    }
                }
            }

            if (!flag) break
            answer++
        }

        sb.append(answer)
        bw.write(sb.toString())

        br.close()
        bw.close()

    }

    val dx = arrayOf(1, -1, 0, 0)
    val dy = arrayOf(0, 0, 1, -1)

    fun bfs(n: Int, l: Int, r: Int, x: Int, y: Int): List<Node> {
        val node = Node(x, y)
        val nodes = mutableListOf<Node>()
        val deque = ArrayDeque<Node>()
        deque.add(node)
        visited[x][y] = true

        while (deque.isNotEmpty()) {
            val (sx, sy) = deque.removeFirst()

            nodes.add(Node(sx, sy))

            repeat(4) { i ->
                val nx = sx + dx[i]
                val ny = sy + dy[i]

                if (nx in 0 until n && ny in 0 until n && !visited[nx][ny]) {
                    if (Math.abs(graph[sx][sy] - graph[nx][ny]) in l..r) {
                        visited[nx][ny] = true
                        deque.add(Node(nx, ny))
                    }
                }
            }
        }
        return nodes
    }

}

fun main() {
    `16234`().solve()
}
